src/agni/server/communication/MessageReceiver.java

```diff
@@ -9,20 +17,196 @@
+    FileReceiver fileReceiver = null;
+    HeartbeatReceiver heartBeatReceiver = null;
+    InfoRequestReceiver infoRequestReceiver = null;
+    LoginReceiver loginReceiver = null;
+    UserReceiver userReceiver = null;
+
     
     public MessageReceiver(ChannelList channels,
-                           LoginReceiver loginReceiver,
-                           UserReceiver userReceiver,
-                           ChatReceiver chatReceiver,
-                           FileReceiver fileReceiver,
-                           InfoRequestReceiver infoRequestReceiver, 
-                           HeartbeatReceiver heartbeatReceiver) {
+            LoginReceiver loginReceiver,
+            UserReceiver userReceiver,
+            ChatReceiver chatReceiver,
+            FileReceiver fileReceiver,
+            InfoRequestReceiver infoRequestReceiver, 
+            HeartbeatReceiver heartbeatReceiver) {
+        
+        this.loginReceiver = loginReceiver;
+        this.userReceiver = userReceiver;
+        this.chatReceiver = chatReceiver;
+        this.fileReceiver = fileReceiver;
+        this.infoRequestReceiver = infoRequestReceiver; 
+        this.heartBeatReceiver = heartbeatReceiver;
+        this.channelList = channels;
+    }
+    
+    public enum MessageTypes {
+        HEARTBEAT((byte)0x01),
+        LOGIN((byte)0x02),
+        INFO((byte)0x03),
+        STATUS((byte)0x04),
+        CHAT((byte)0x05),
+        FILE((byte)0x06);
+
+        private final byte bytes;
+        private MessageTypes(byte bytes) {
+            this.bytes = bytes;
+        }
+         final public byte bytes() {
+            return bytes;
+         }
+      }
+
+/*
+ *@requires server port number
+ *@promises selector initialized and ready for connection requests
+ *@promises socket channel is initialized to be non-blocking and bound to specified port
+ */
+    public void initializeConnection(String port) {
+        try {
+            ServerSocketChannel.open();
+            channel.configureBlocking(false);
+            InetSocketAddress isa = new InetSocketAddress (Integer.parseInt(port));
+            channel.socket().bind(isa);
+
+            selector = Selector.open();
+            channel.register(selector, SelectionKey.OP_ACCEPT);
+        } catch (IOException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
+    }
+    
+    private void saveIpChannelPair(SocketChannel channel) {
+        String ip = null;
+        try {
+            ip = channel.getRemoteAddress().toString();
+        } catch (IOException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
+        channelList.addPair(ip, channel);
+    }
+    
+/*
+ * @requires ip of current channel
+ * @requires message buffer
+ * @promises call appropriate 'receiverMessage(ip, buffer)' method
+ */
+    private void selectReceiver(String ip, ByteBuffer buffer){
+        byte messageType = buffer.get(4);
```

Comment:
This will increment the position, so the buffer passed into the argument here won't be the same one that we're passing in to the Receivers. I think this will become a problem.